### DISCUSSION TITLE
Add Nibble/DSK Image support

### DISCUSSION THREAD

**Discussion by bfranske** [tjboldt/Apple2-IO-RPi#71] [General]
**Is your feature request related to a problem? Please describe.**
Depends on how you define problem :) The problem is Apple2-IO-RPi does not currently support Nibble/DSK images which are useful for software which can't be easily loaded onto a ProDOS volume.

**Describe the solution you'd like**
Implement Nibble/DSK support for DOS 3.3 images, see suggestions on method below.

**Additional context**
I'm a bit naive with Apple II design...but I think this is doable with software/firmware only changes? Did some research and it looks like an approach similar to the open source [Pseudo-Disk II](http://alexfreed.com/FPGApple/pseudodisk.html) would be possible. That project supports both a ProDOS and Nibble mode. The general approach to Nibble support is described as "Boot code identical to one found in a standard DISK II controller card is mapped into the Cx00-CxFF memory space. AVR interprets reads from the addresses responsible for disk operation. For example it notices attempts to position the head and adds a offset to the beginning of the image." In the case of the Apple2-IO-RPi it would seem that the Boot ROM could load the Disk II card code into memory, then a driver on the Pi side (instead of the AVR used by Pseudo-Disk II) could interpret the calls for disk operation and read/write from a Nibble (or DSK) file.

**Comment by tjboldt** [comment]
The Raspberry Pi running Linux is not particularly well suited for realtime. The Pseudo-Disk II concept really requires different hardware with a microcontroller. This request would be a good candidate for a Raspberry Pi Pico but I believe someone else is already working on a project for that. This project was designed for command and response, the Apple II sending commands and the Raspberry Pi responding. The focus going forward is to increase what those commands are from the existing ReadBlock, WriteBlock, Execute, Shell, LoadFile etc. and add on things like LoadImage, Compile, WGet, VNC, etc.

**Comment by tjboldt** [comment]
Moved to Discussions

**Reply by bfranske** [reply]
FWIW I started looking into this further myself... Upon closer examination of the schematics it seems like this is not possible with the current hardware revision. In particular, unless I've missed something A0, A1, and A2 are needed for detecting the pseudo head location but those are not actually available to the Pi in the A2-IO-RPi. I do suspect that the Pi would be fast enough to directly emulate the Disk II controller without needing CPLD glue logic as the Pseudo Disk II did with it's AVR chip.

If I am going to have to design and spin my own PCB though this did get me thinking about what I would really, really want... and that is to use the Pi as a general purpose "software re-programmable Apple II expansion card" including the ability to do DMA on the Apple 2. This would allow it to be a platform which people with software skills, who lack hardware skills, could use to develop interesting Apple 2 cards entirely in software. To really do that though you need all the address lines and data lines available to the Pi. That's more I/O lines than the Pi has available...

So, then I started thinking if I need to do GPIO expansion (e.g. with an SPI or I2C I/O chip) why not instead do it with a RP2040 controller... That would allow for IRQs on all the pins in case there are ultra time-sensitive things that people might want to do. So my current plan is to design a PCB which basically is nothing more than an EEPROM, and connection via level shifters for all the address and data bus lines from the Apple 2 (plus a few others of import) to the RP2040 which then communicates to an outboard Pi for the real processing. That connection is probably to be over USB for simplicity. I also want to make the EEPROM re-programmable from the Pi via the Pico so that the whole card can be flexibly reprogrammed on demand.

Ideally the Pi would be able to emulate a variety of cards/peripherals including: Disk II, SmartPort/ProDos HDD, Serial, Parallel, Apple II Workstation (EtherTalk w/ network booting), Printers, etc. as well as new ideas for experimentation and virtual cards people might have.

**Reply by tjboldt** [reply]
It sounds like an interesting and ambitious project. I would love to hear about your progress. The GPIO on the RPi Z W 2 is almost fast enough to directly interact with the Apple II bus but Linux is not an RTOS so you will get times where data is lost when the OS is busy unless you add latches and other chips. The RP2040 is definitely fast enough and real time but it lacks wifi connectivity and storage without adding more support components. A combination of both RP2040 and RPI would be amazing but might be a lot of work for a hobby.

**Comment by bfranske** [comment]
As it relates specifically to emulating a Disk II card and drive. My understanding is that a Pi directly would probably be able to keep up with the actual data transfer in and out since the data is latched and the Pi runs so fast comparatively. The speed concern would mostly come as to whether the Pi would miss the stepper motor track movement on the A0-A3 lines which is possible. My gut instinct is that it may be fast enough directly but... for an extra $4 it seems like both good insurance and a way to substantially increase the potential uses for experimentation.
